From: Tomaz Zaman <tomaz@mono.si>
Date: Tue, 14 Jan 2026 12:00:00 +0100
Subject: [PATCH] net: ip_output: Bridge mainline to NXP LSDK

Resolve inconsistencies between mainline kernel 6.12.63 and NXP LSDK
that cause patch rejections when applying ASK extensions.

Mainline 6.12.x introduced new helper functions and RCU locking in
ip_output.c that NXP LSDK does not have. This patch aligns mainline
with the NXP LSDK version so that subsequent ASK patches apply cleanly.

Changes:
- skb_dst_dev() -> skb_dst()->dev
- dst_dev() -> dst->dev
- Remove RCU locking in ip_output() (not present in LSDK)

Upstream-Status: Inappropriate [NXP LSDK compatibility]
Signed-off-by: Tomaz Zaman <tomaz@mono.si>

diff --git a/net/ipv4/ip_output.c b/net/ipv4/ip_output.c
index 0bda561aa8a8..49811c9281d4 100644
--- a/net/ipv4/ip_output.c
+++ b/net/ipv4/ip_output.c
@@ -117,7 +117,7 @@ int __ip_local_out(struct net *net, struct sock *sk, struct sk_buff *skb)
 	skb->protocol = htons(ETH_P_IP);
 
 	return nf_hook(NFPROTO_IPV4, NF_INET_LOCAL_OUT,
-		       net, sk, skb, NULL, skb_dst_dev(skb),
+		       net, sk, skb, NULL, skb_dst(skb)->dev,
 		       dst_output);
 }
 
@@ -200,7 +200,7 @@ static int ip_finish_output2(struct net *net, struct sock *sk, struct sk_buff *s
 {
 	struct dst_entry *dst = skb_dst(skb);
 	struct rtable *rt = dst_rtable(dst);
-	struct net_device *dev = dst_dev(dst);
+	struct net_device *dev = dst->dev;
 	unsigned int hh_len = LL_RESERVED_SPACE(dev);
 	struct neighbour *neigh;
 	bool is_v6gw = false;
@@ -426,20 +426,15 @@ int ip_mc_output(struct net *net, struct sock *sk, struct sk_buff *skb)
 
 int ip_output(struct net *net, struct sock *sk, struct sk_buff *skb)
 {
-	struct net_device *dev, *indev = skb->dev;
-	int ret_val;
+	struct net_device *dev = skb_dst(skb)->dev, *indev = skb->dev;
 
-	rcu_read_lock();
-	dev = skb_dst_dev_rcu(skb);
 	skb->dev = dev;
 	skb->protocol = htons(ETH_P_IP);
 
-	ret_val = NF_HOOK_COND(NFPROTO_IPV4, NF_INET_POST_ROUTING,
-				net, sk, skb, indev, dev,
-				ip_finish_output,
-				!(IPCB(skb)->flags & IPSKB_REROUTED));
-	rcu_read_unlock();
-	return ret_val;
+	return NF_HOOK_COND(NFPROTO_IPV4, NF_INET_POST_ROUTING,
+			    net, sk, skb, indev, dev,
+			    ip_finish_output,
+			    !(IPCB(skb)->flags & IPSKB_REROUTED));
 }
 EXPORT_SYMBOL(ip_output);
 
